firmware/powerboard2/code/app/powerManagement.c

```diff
@@ -0,0 +1,212 @@
+            //powerManagement_data.batteryCurrents[POWER_MANAGEMENT_BATTERY_CHANNEL_RIGHT] = ADC_VALUE_TO_CURRENT(ADC_getChannelData(ADC_CHANNEL_RIGHT_BATT_CURRENT));
+
+            powerManagement_setState(POWER_MANAGEMENT_CHANNEL_VBATT, message.TRIDENT_powerEnable.VBattPowerEnable);
+            powerManagement_setState(POWER_MANAGEMENT_CHANNEL_5V, message.TRIDENT_powerEnable._5VPowerEnable);
+            powerManagement_setState(POWER_MANAGEMENT_CHANNEL_12V, message.TRIDENT_powerEnable._12VPowerEnable);
+            powerManagement_setState(POWER_MANAGEMENT_CHANNEL_16V, message.TRIDENT_powerEnable._16VPowerEnable);
+			
+			/*Implement this after housingMonitor code is done
+			//Ckeck if leakage happens in the main housing before turning on power
+			housingMonitor_housing_E Main_housing = HOUSINGMONITOR_HOUSING_MAIN;
+			if (housingMonitor_getHousingStatus(Main_housing) == HOUSINGMONITOR_HOUSINGSTATUS_LEAK)
+			{
+				powerManagement_setState(POWER_MANAGEMENT_CHANNEL_MOTOR, false);
+				powerManagement_setState(POWER_MANAGEMENT_CHANNEL_5V, false);
+				powerManagement_setState(POWER_MANAGEMENT_CHANNEL_12V_9V, false);
+			}
+			//No leakage. If power enable is true, turn on the power.
+			else 
+			{
+				powerManagement_setState(POWER_MANAGEMENT_CHANNEL_MOTOR, message.POLARIS_powerEnable.motorPowerEnable);
+				powerManagement_setState(POWER_MANAGEMENT_CHANNEL_5V, message.POLARIS_powerEnable._5VPowerEnable);
+				powerManagement_setState(POWER_MANAGEMENT_CHANNEL_12V_9V, message.POLARIS_powerEnable._12V9VPowerEnable);
+			}*/
+
+            break;
+        }
+        default:
+        {
+            break;
+        }
+    }
+}
+
+void powerManagement_setState(const powerManagement_channel_E channel, const bool newState)
+{
+    if(channel < POWER_MANAGEMENT_CHANNEL_COUNT)
+    {
+        GPIO_TypeDef * GPIOPort = GPIOC;
+        uint16_t GPIOPin = 0;
+
+        switch(channel)
+        {
+            case POWER_MANAGEMENT_CHANNEL_VBATT:
+            {
+                //GPIOPort = GPIOC;
+                GPIOPin = _VBATT_ENABLE_GPIO;
+                break;
+            }
+            case POWER_MANAGEMENT_CHANNEL_5V:
+            {
+                //GPIOPort = GPIOA;
+                GPIOPin = _5V_ENABLE_GPIO;
+                break;
+            }
+            case POWER_MANAGEMENT_CHANNEL_12V:
+            {
+                //GPIOPort = GPIOC;
+                GPIOPin = _12V_ENABLE_GPIO;
+                break;
+            }
+            case POWER_MANAGEMENT_CHANNEL_16V:
+            {
+                //GPIOPort = GPIOC;
+                GPIOPin = _16V_ENABLE_GPIO;
+                break;
+            }
+            case POWER_MANAGEMENT_CHANNEL_COUNT:
+            default:
+            {
+                break;
+            }
+        }
+
+        if(GPIOPin != 0)
+        {
+            if(newState == true)
+            {
+                GPIOPort->ODR |= GPIOPin;
```

Comment:
Just writing to the GPIO pin won't turn on the power supply. There is a shift register operating in parallel mode inbetween. You need to set the GPIO pin, then pulse the pin connected to the clock on the shit register.